File: app/gui-electron/src/fsOps.js
/**
 * fsOps - abstraction layer for file & script operations
 * - readFile, writeFile, applyUnifiedDiff (lightweight), runScript
 * - returns normalized results for the renderer
 */
const fs = require('fs').promises;
const path = require('path');
const { spawn } = require('child_process');
const os = require('os');
const crypto = require('crypto');

function makeId(prefix='id') {
  if (crypto.randomUUID) return `${prefix}-${crypto.randomUUID()}`;
  return `${prefix}-${Date.now()}-${Math.floor(Math.random()*1e6)}`;
}

async function readTextFile(filePath, { maxBytes=64*1024 } = {}) {
  const content = await fs.readFile(filePath, 'utf8');
  if (content.length > maxBytes) {
    return content.slice(0, maxBytes/2) + '\n...TRUNCATED...\n' + content.slice(-maxBytes/2);
  }
  return content;
}

async function writeTextFile(filePath, content, { ensureDir=true, mode } = {}) {
  if (ensureDir) {
    await fs.mkdir(path.dirname(filePath), { recursive: true });
  }
  await fs.writeFile(filePath, content, 'utf8');
  if (mode) await fs.chmod(filePath, mode);
  return { ok: true, path: filePath };
}

// Very small, safe unified-diff applier for simple cases (ctx: small project patches).
// NOTE: This is a best-effort tool — complex diffs should use an external patcher.
async function applyUnifiedDiff(filePath, unifiedDiff, { dryRun=false } = {}) {
  // Simple fallback: if diff starts with "+++ " and "@@ " we attempt to replace entire file if diff contains "===REPLACE===\n<content>"
  // Safer approach: if unifiedDiff contains a marker "===REPLACE===", treat payload after marker as replacement content.
  const REPLACE_MARKER = '===REPLACE===';
  if (unifiedDiff.includes(REPLACE_MARKER)) {
    const parts = unifiedDiff.split(REPLACE_MARKER);
    const newContent = parts[1];
    if (dryRun) return { applied: true, reason: 'dry-run: replace detected' };
    await writeTextFile(filePath, newContent);
    return { applied: true, file: filePath };
  }
  // otherwise, no-op (caller should fall back to manual patch).
  return { applied: false, reason: 'unsupported-diff-format' };
}

function runScript(command, args = [], opts = {}) {
  const cwd = opts.cwd || process.cwd();
  const timeoutMs = opts.timeoutMs || 60_000;
  return new Promise((resolve) => {
    const child = spawn(command, args, { cwd, shell: true, env: process.env });
    let stdout = '';
    let stderr = '';
    const to = setTimeout(() => {
      child.kill('SIGKILL');
    }, timeoutMs);

    child.stdout?.on('data', (d) => { stdout += d.toString(); });
    child.stderr?.on('data', (d) => { stderr += d.toString(); });
    child.on('error', (err) => {
      clearTimeout(to);
      resolve({ code: null, error: err.message, stdout, stderr });
    });
    child.on('close', (code, signal) => {
      clearTimeout(to);
      resolve({ code, signal, stdout, stderr });
    });
  });
}

module.exports = {
  readTextFile,
  writeTextFile,
  applyUnifiedDiff,
  runScript,
  makeId
};

File: warp_core/src/fs_ops.rs
//! fsOps - abstraction layer for file & script operations
//! - read_text_file, write_text_file, apply_unified_diff, run_script
//! - returns normalized results matching Phase 4 JS API

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::process::Stdio;
use tokio::fs;
use tokio::process::Command;
use tokio::time::{timeout, Duration};

#[derive(Debug, Serialize, Deserialize)]
pub struct ReadFileOpts {
    pub max_bytes: usize,
}

impl Default for ReadFileOpts {
    fn default() -> Self {
        Self {
            max_bytes: 64 * 1024,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WriteFileOpts {
    pub ensure_dir: bool,
    pub mode: Option<u32>,
}

impl Default for WriteFileOpts {
    fn default() -> Self {
        Self {
            ensure_dir: true,
            mode: None,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WriteFileResult {
    pub ok: bool,
    pub path: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ApplyDiffOpts {
    pub dry_run: bool,
}

impl Default for ApplyDiffOpts {
    fn default() -> Self {
        Self { dry_run: false }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ApplyDiffResult {
    pub applied: bool,
    pub reason: Option<String>,
    pub file: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RunScriptOpts {
    pub cwd: Option<String>,
    pub timeout_ms: Option<u64>,
}

impl Default for RunScriptOpts {
    fn default() -> Self {
        Self {
            cwd: None,
            timeout_ms: Some(60_000),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RunScriptResult {
    pub code: Option<i32>,
    pub signal: Option<String>,
    pub stdout: String,
    pub stderr: String,
    pub error: Option<String>,
}

/// Generate a unique ID with prefix (matching JS makeId)
pub fn make_id(prefix: &str) -> String {
    format!("{}-{}", prefix, uuid::Uuid::new_v4())
}

/// Read a text file with optional truncation for large files
pub async fn read_text_file(file_path: impl AsRef<Path>, opts: ReadFileOpts) -> Result<String> {
    let path = file_path.as_ref();
    let content = fs::read_to_string(path)
        .await
        .context(format!("Failed to read file: {:?}", path))?;

    if content.len() > opts.max_bytes {
        let half = opts.max_bytes / 2;
        let start = &content[..half];
        let end = &content[content.len() - half..];
        Ok(format!("{}\n...TRUNCATED...\n{}", start, end))
    } else {
        Ok(content)
    }
}

/// Write a text file with optional directory creation and permissions
pub async fn write_text_file(
    file_path: impl AsRef<Path>,
    content: &str,
    opts: WriteFileOpts,
) -> Result<WriteFileResult> {
    let path = file_path.as_ref();

    if opts.ensure_dir {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)
                .await
                .context("Failed to create parent directory")?;
        }
    }

    fs::write(path, content)
        .await
        .context(format!("Failed to write file: {:?}", path))?;

    #[cfg(unix)]
    if let Some(mode) = opts.mode {
        use std::os::unix::fs::PermissionsExt;
        let permissions = std::fs::Permissions::from_mode(mode);
        std::fs::set_permissions(path, permissions)
            .context("Failed to set file permissions")?;
    }

    Ok(WriteFileResult {
        ok: true,
        path: path.display().to_string(),
    })
}

/// Apply a unified diff with simple replacement marker support
/// Matches JS behavior: if diff contains "===REPLACE===" marker, 
/// treat content after marker as full file replacement
pub async fn apply_unified_diff(
    file_path: impl AsRef<Path>,
    unified_diff: &str,
    opts: ApplyDiffOpts,
) -> Result<ApplyDiffResult> {
    const REPLACE_MARKER: &str = "===REPLACE===";

    if let Some(marker_pos) = unified_diff.find(REPLACE_MARKER) {
        let new_content = &unified_diff[marker_pos + REPLACE_MARKER.len()..];

        if opts.dry_run {
            return Ok(ApplyDiffResult {
                applied: true,
                reason: Some("dry-run: replace detected".to_string()),
                file: None,
            });
        }

        let path = file_path.as_ref();
        write_text_file(path, new_content, WriteFileOpts::default()).await?;

        return Ok(ApplyDiffResult {
            applied: true,
            reason: None,
            file: Some(path.display().to_string()),
        });
    }

    // No marker found - return unsupported
    Ok(ApplyDiffResult {
        applied: false,
        reason: Some("unsupported-diff-format".to_string()),
        file: None,
    })
}

/// Run a shell script/command with timeout
pub async fn run_script(
    command: &str,
    args: Vec<String>,
    opts: RunScriptOpts,
) -> Result<RunScriptResult> {
    let cwd = opts
        .cwd
        .as_ref()
        .map(PathBuf::from)
        .unwrap_or_else(|| std::env::current_dir().unwrap_or_default());

    let timeout_duration = Duration::from_millis(opts.timeout_ms.unwrap_or(60_000));

    let mut cmd = Command::new(command);
    cmd.args(&args)
        .current_dir(cwd)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    match cmd.spawn() {
        Ok(child) => {
            // Use async block to handle timeout without moving child
            let result = timeout(timeout_duration, async {
                let output = child.wait_with_output().await?;
                Ok::<_, std::io::Error>(output)
            }).await;

            match result {
                Ok(Ok(output)) => Ok(RunScriptResult {
                    code: output.status.code(),
                    signal: None,
                    stdout: String::from_utf8_lossy(&output.stdout).to_string(),
                    stderr: String::from_utf8_lossy(&output.stderr).to_string(),
                    error: None,
                }),
                Ok(Err(e)) => Ok(RunScriptResult {
                    code: None,
                    signal: None,
                    stdout: String::new(),
                    stderr: String::new(),
                    error: Some(e.to_string()),
                }),
                Err(_) => {
                    // Timeout exceeded
                    Ok(RunScriptResult {
                        code: None,
                        signal: Some("SIGKILL".to_string()),
                        stdout: String::new(),
                        stderr: String::new(),
                        error: Some("Timeout exceeded".to_string()),
                    })
                }
            }
        }
        Err(e) => Ok(RunScriptResult {
            code: None,
            signal: None,
            stdout: String::new(),
            stderr: String::new(),
            error: Some(e.to_string()),
        }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_make_id() {
        let id = make_id("test");
        assert!(id.starts_with("test-"));
    }

    #[tokio::test]
    async fn test_write_and_read() {
        let temp_dir = std::env::temp_dir();
        let test_file = temp_dir.join("warp_core_test.txt");
        let content = "Hello Warp Core!";

        let result = write_text_file(&test_file, content, WriteFileOpts::default())
            .await
            .unwrap();
        assert!(result.ok);

        let read_content = read_text_file(&test_file, ReadFileOpts::default())
            .await
            .unwrap();
        assert_eq!(read_content, content);

        let _ = fs::remove_file(&test_file).await;
    }
}
